soc: raspberrypi: rp2350: support AMP builds

dsseng opened the pull request:
Continuation of #90922

To test:

```bash
west build -d build/0 -b rpi_pico2/rp2350a/m33/cpu0 samples/basic/blinky --pristine=auto -- -DCONFIG_XIP=n -DCONFIG_SOC_RP2350_CPU1_ENABLE=y -DCONFIG_SOC_LOG_LEVEL_DBG=y && west build -d build/1 -b rpi_pico2/rp2350a/m33/cpu1 samples/subsys/usb/webusb --pristine=auto && picotool load -nuv -t elf ./build/1/zephyr/zephyr.elf && picotool load -nuvx -t elf ./build/0/zephyr/zephyr.elf
```

Or see https://github.com/dsseng/zephyr/commits/dss-worktree/ for integration tests with IPC

dancollins committed:
soc: rp2350: skip clock initialisation for cpu1

We don't want to reinitialise the clock control when
we boot the second core - so this commit introduces a
means to skip initialisation (i.e. assume it is already
initialised).

Co-authored-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>
Signed-off-by: Dan Collins <dan@collinsnz.com>
Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
dts: raspberrypi: rpi_pico-led: disable by default

Will only enable in core-specific dts.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
dts: arm: rpi_pico2: do not enable peripherals in the common dtsi

To ensure AMP works, only enable peripherals in the final dts

Co-authored-by: Dan Collins <dan@collinsnz.com>
Signed-off-by: Dan Collins <dan@collinsnz.com>
Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dancollins committed:
boards: rpi_pico2: create CPU0/1 targets

CPU0 gets flash, UART0, 256KB SRAM

CPU1 gets USB, 264KB SRAM (RAM boot only for now)

This commit introduces a means to build Zephyr for both
Cortex M33 cores of the RP2350, and for CPU0 to boot CPU1.
It includes Kconfig changes to build for both cores in an AMP
configuration.

Co-authored-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>
Signed-off-by: Dan Collins <dan@collinsnz.com>
Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
samples: ipc: icmsg: support RP2350

Add configs for interprocessor communication on rpi_pico2

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
soc: rp2350: add support for building for CPU1 with a code partition

This change enables building an image to be stored in a flash partition
to be loaded using `image-source`.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented:
> Hello, I’m interested in this feature; if you need, I can test it on the RP2040. Kind regards, Thomas Perrot

Hello Thomas, I'm glad you offered your help to improve this work. You can see the current AMP work assembled for the RP2350 here: https://github.com/dsseng/zephyr/commits/dss-worktree/. You will need some devicetree and Kconfig knowledge to port all of that to RP2040. What board are you using?

If possible, reach out to the `#raspberrypi` Discord channel, as it'd be easier to cooperate in a chat mode. Thanks!

dsseng commented on dts/bindings/misc/raspberrypi,rp2-cpu1.yaml:
```diff
@@ -0,0 +1,39 @@
+  Secondary CPU launcher for Raspberry Pi RP2 SoCs.
+
+  RP2040 and RP2350 SoCs have a secondary CPU (CPU1) that can be used
+  for launching AMP applications. This driver can load the CPU1 image
```
I modeled the driver based on VPR secondary core handling for nRF54 SoCs. Those have similar bindings, also for misc, and not CPU. Wording can be improved, agreed.

dsseng replied:
Hm, nRF VPR does have a launcher DT node. I applied the rule of following an equivalent which is already in tree. However, this might of course be not the single proper way. Could we maybe discuss that in the `#raspberrypi` Discord channel, considering all pros and cons of both keeping it in a driver (specifying partitions and memory via DT) or moving to SoC (specifying memory regions using Kconfig)

dancollins committed:
soc: raspberrypi: rp2350: support booting CPU1

When configured to do so, the SoC init code will load the CPU1 app,
reset CPU1 and boot it into the app.

There is also an optional validation method to make sure CPU1 VTOR is
valid.

Modeled similarly to soc/nordic/nrf54h/soc.c

Co-authored-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>
Signed-off-by: Dan Collins <dan@collinsnz.com>
Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng committed:
tests: ipc: ipc_sessions: run on rpi_pico2

Disable reboot since CPU1 cannot reset itself.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

dsseng commented on soc/raspberrypi/rpi_pico/rp2350/soc.c:
```diff
@@ -22,3 +24,182 @@ void soc_reset_hook(void)
+
+LOG_MODULE_REGISTER(soc_rp2350, CONFIG_SOC_LOG_LEVEL);
+
+#if !DT_NODE_EXISTS(DT_NODELABEL(sram0_cpu1))
```
Nice catch, done, thank you!

dsseng commented on boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_cpu0.yaml:
```diff
@@ -0,0 +1,21 @@
+  - zephyr
+  - gnuarmemb
+supported:
+  - adc
+  - clock
+  - counter
+  - dma
+  - gpio
+  - hwinfo
+  - i2c
+  - pwm
+  - spi
+  - uart
```
I know. You're right, it probably makes sense to just drop these tests, since they replicate the plain `cm33` variant. Thanks

soburi commented on boards/raspberrypi/common/rpi_pico-pinctrl-common.dtsi:
```diff
@@ -15,6 +15,16 @@
 		};
 	};
 
+	uart1_default: uart1_default {
```
This is a request because rpi_pico has a shield definition, 
https://github.com/zephyrproject-rtos/zephyr/blob/1b97f4b2b09cc940a2eb56cc869a7bb26fb35c32/boards/raspberrypi/rpi_pico2/rpi_pico2.dtsi#L28-L59

It is a consideration to ensure that only the default settings need to be considered when creating a shield definition.
https://datasheets.raspberrypi.com/pico/Pico-2-Pinout.pdf

dsseng commented on soc/raspberrypi/rpi_pico/soc.yml:
```diff
@@ -9,6 +9,8 @@ family:
     - name: rp2350a
       cpuclusters:
       - name: m33
```
No, I'm keeping the "whole board under one CPU" scenario. Later this could use cores in a symmetric multiprocessing configuration.

nordicjm replied:
it is not a SoC thing but a board thing so `rpi_pico2/rp2350a/m33` would be by itself, or AMP (if you build 2 images, maybe with `rpi_pico2/rp2350a/m33/seccpu` or `rpi_pico2/rp2350a/m33/cpu1` or something else for the second image), then SMP would be `rpi_pico2/rp2350a/m33/smp`. Example with SMP: https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/snps/nsim/arc_classic
Example with 1 core with 2 "images" (TF-M): https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/arm/mps2/

afantino951 commented on soc/raspberrypi/rpi_pico/rp2350/Kconfig:
```diff
@@ -30,10 +50,46 @@ config SOC_RP2350B_M33
 config RP2_REQUIRES_IMAGE_DEFINITION_BLOCK
 	bool
-	default y
+	# IMAGE_DEF should be used to define the image for booting CPU0
+	default y if !SOC_RP2350A_M33_CPU1
```
When building the test command with `zephyr,code-partition` defined at SRAM addr `0x20020000`, I noticed that the build output for the test command's CPU1 code contains the following excerpt:

```
Memory region         Used Size  Region Size  %age Used
           FLASH:          0 GB       128 KB      0.00%
             RAM:       60292 B       256 KB     23.00%
        IDT_LIST:          0 GB        32 KB      0.00%
Generating files from /home/fantino/oss-afa/zafa-fork/zephyr/build/1/zephyr/zephyr.elf for board: rpi_pico2
Converted to uf2, output size: 95232, start address: 0x40040000
```

The uf2's start address is `0x40040000` instead of `0x20020000`. I think this is because:

```
SOC_RP2350A_M33_CPU1=y
  -> RP2_REQUIRES_IMAGE_DEFINITION_BLOCK=n
    -> BUILD_OUTPUT_UF2_USE_FLASH_BASE=n 
    -> BUILD_OUTPUT_UF2_USE_FLASH_OFFSET=y
```

The `CONFIG_BUILD_OUTPUT_*` options are used here:

https://github.com/zephyrproject-rtos/zephyr/blob/ca68c771988a84e522c631182f08388663dd3b01/CMakeLists.txt#L1883-L1897

and `FLASH_LOAD_OFFSET` is defined here:

https://github.com/zephyrproject-rtos/zephyr/blob/ca68c771988a84e522c631182f08388663dd3b01/Kconfig.zephyr#L100-L113

Where we would essentially do `code_addr = CONFIG_FLASH_LOAD_OFFSET + CONFIG_FLASH_LOAD_OFFSET`, which is why we see the uf2 start address doubled.

dsseng replied:
I guess I've done a fix for that, #94554 . Not sure if this branch or `dss-worktree` contains this fix though.

For testing I uploaded ELF files mostly.

Thanks for review, will take a look

dsseng commented:
Force-push: rebased, addressed review comments via fixups

nordicjm commented on soc/raspberrypi/rpi_pico/rp2350/Kconfig:
```diff
@@ -79,6 +79,7 @@ config SOC_RP2350B_M33
 	select CPU_HAS_FPU
 	select ARMV8_M_DSP
 
+if SOC_SERIES_RP2350
```
newline gap spacing missing

nordicjm commented on boards/raspberrypi/rpi_pico2/board.yml:
```diff
@@ -12,3 +12,7 @@ board:
               cpucluster: m33
         - name: mcuboot
           cpucluster: m33
+        - name: cpu0
```
PR is a bit confusing and needs rework, the end solution looks to be OK, but an early commit adds extra things to the soc file, hence why I was confused, because later on they are entirely removed and replacing with board changes (board changes are fine) so the soc.yml etc. changes can go

dsseng replied:
@nordicjm could you please re-review? The commit history should be reviewable as well, and all your previous comments have been addressed

JarmouniA commented on soc/raspberrypi/rpi_pico/rp2350/Kconfig:
```diff
@@ -57,10 +79,32 @@ config SOC_RP2350B_M33
+	  entrypoint during SoC late init hook.
+
+config SOC_RP2350_CPU1_ENABLE_CHECK_VTOR
+	bool "Check VTOR before booting the second core"
```
Does this need to be user-selectable?

JarmouniA commented on soc/raspberrypi/rpi_pico/rp2350/soc.c:
```diff
@@ -29,3 +31,176 @@ void soc_reset_hook(void)
+	return sio_regs->fifo_rd;
+}
+
+#if DT_NODE_EXISTS(DT_NODELABEL(cpu1_slot0_partition))
```
You can get a partition node by label using DT_NODE_BY_FIXED_PARTITION_LABEL

dsseng commented on boards/raspberrypi/rpi_pico2/rpi_pico2_rp2350a_m33_w.dts:
```diff
@@ -21,6 +21,67 @@
  */
 #include "rpi_pico2.dtsi"
```
Thanks for pointing this out. Actually, it's led0 that has to be removed, since it is not connected to the SoC on W series, but rather to the wireless chip

afantino951 replied:
Oh that's right! Thanks for the explanation!

dsseng committed:
drivers: mbox: rpi_pico: factor out low-level access functions

Move them into a header for sharing with SoC code for booting CPU1.

Signed-off-by: Dmitrii Sharshakov <d3dx12.xx@gmail.com>

andylinpersonal commented on soc/raspberrypi/rpi_pico/rp2350/soc.c:
```diff
@@ -29,3 +31,176 @@ void soc_reset_hook(void)
+	sio_regs->fifo_wr = value;
+
+	/* Inform other CPU about FIFO update. */
+	__SEV();
```
The conflicting functions: `__sev`, `__wfe` and `__wfi` have already been guarded with `__has_builtin`. Did I miss something else?

afantino951 committed:
dts: arm: rpi_pico2: Remove unsupported UART from cpu1 board

The rpi_pico2/rp2350a/m33/cpu1 board does not have a uart device enabled
and does not set /chosen/zephyr,console and /chosen/zephyr,shell-uart.
This would lead to warnings in the build process for the samples. Remove
uart from the board's supported list and remove the Kconfig items that
enable the console and uart.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
soc: rp2350: Use generic partition macros in cpu1 launch code

Update cpu1 launch code to use the new DT_PARTITION_* macros instead of
the the DT_FIXED_PARTITION_* macros to allow for cpu1_slot0_partition to
be either a zephyr,mapped-partition or a fixed-partition.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
boards: rpi_pico2: Move rpi_pico2_split to vendor directory

Moved boards/.../rpi_pico2_split.dtsi to
dts/vendor/.../rp2350_shared_reserved_memory.dtsi so it can be imported
by snippets and other rp2350-based boards.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
dts: rpi_pico2: Migrate fixed-partition to zephyr,mapped-partition

Migrate to the new zephyr,mapped-partition flash partition. Memory
mapped partitions should now use the zephyr,mapped-partition compat to
avoid a build warning.

This commit does the minimum changes to enable rp2350 AMP work to
continue, and this work will be expanded on and added to a separate
PR.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
snippets: raspberrypi: Add snippet to launch cpu1

Add snippet that allows users to launch CPU1 from CPU0 for the
rpi_pico2. This is modeled after the nordic-flpr snippet.

Note: The CPU1 image must be built with CONFIG_XIP=n

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
boards: rpi_pico2: Set cpu1 code partition in flash addr space

Set cpu1's zephyr,code-partition to within the non-volatile memory
address space. This is necessary to have our image persist between power
cycles.

I am not sure if it is ok to have /chosen/zephyr,code-partition in an
address space that is not "owned" by cpu1, but I am only using this to
get the right start address for the binaries. Once cpu1 is launched from
cpu0, I think cpu1 should be running entirely from &sram_cpu1.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
samples: ipc: icmsg: Support rp2350/m33 board targets

Support the icmsg sample for the rpi_pico/rp2350a/m33 board. Note that
there is no cpu0 board variant. This is similar to the m33/cpu0 sample
target, but instead of the board variant, it uses the rp2xxx-cpu1
snippet to configure the board to launch the secondary core.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
tests: ipc: ipc_sesions: Move rpi_pico tests to run with snippet

Migrate from the custom board variant, rpi_pico2/rp2350a/m33/cpu0, to
the generic board, rpi_pico2/rp2350a/m33, plus the rp2xxx-cpu1 snippet.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
samples: ipc: icmsg: Remove rp2350a/m33/cpu0 samples

Remove the rpi_pico2/rp2350a/m33/cpu0 board variant icmsg samples. These
are replaced by the rpi_pico2/rp2350a/m33 board with the rp2xxx-cpu1
snippet.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 committed:
boards: rpi_pico2: Remove CPU0 board variant

Remove the CPU0 board variant for the rpi_pico2 board. Instead of the
custom board variant, we use the rp2xxx-cpu1 snippet on the generic
rpi_pico2/rp2350a/m33 board to launch the secondary core.

Signed-off-by: Andrew Fantino <fantino@qti.qualcomm.com>

afantino951 reviewed (commented):
@nordicjm, I tried to implement what you suggested with the snippet based on the flpr, but I am not sure how to treat the `&flash0` dt node for cpu1. The flpr has its own address space and the main core has a register mapped launcher device. The rp2350 has neither. 

- Should cpu1 "own" the address space it is flashed to even though it shouldn't use it at runtime? 
- Does cpu1 need any flash at all?

dsseng commented:
@nordicjm @soburi - could you please take another look (regarding the code, not commits - will reorder those later)? Thank you!

nordicjm commented:
> @nordicjm @soburi - could you please take another look (regarding the code, not commits - will reorder those later)? Thank you!

I go through PRs one commit at a time so if there's issues with that then please fix before asking for review

dsseng commented:
Deprecated in favor of #109953 as I don't currently have time to finalize this
